### ISSUE TITLE
To many shared libraries linked with a OpenDDS client

### ISSUE THREAD

**Issue by jwillemsen** [objectcomputing/OpenDDS#2094]
When I look for example at the tests/DCPS/ReadCondition I see that this OpenDDS client links with ` -lOpenDDS_Shmem -lOpenDDS_Rtps_Udp -lOpenDDS_Rtps -lOpenDDS_Multicast -lOpenDDS_Udp -lOpenDDS_Tcp -lOpenDDS_InfoRepoDiscovery -lOpenDDS_Dcps -lTAO_BiDirGIOP -lTAO_PI -lTAO_CodecFactory -lTAO_PortableServer -lTAO_AnyTypeCode -lTAO -lACE`

I have the feeling that this is way to much. OpenDDS internally uses portableserver when the info repo is used, but that is all internal to OpenDDS so why should a user application link with all those TAO libraries? Shouldn't for a user application only TAO and ACE be enough? BiDir, PI, CodecFactory, PortableServer and AnyTypeCode are not used by any user code?

**Comment by mitza-oci** [comment]
All libraries that are used (even indirectly) will be listed to support static builds.  There is no downside to including them in shared builds since they need to be present for deployment anyway.

**Comment by mitza-oci** [comment]
As a user you are free to build with OpenDDS_InfoRepoDiscovery and the list will be smaller.  The test is set up to be more flexible so that different configurations can be selected at runtime.

**Comment by jwillemsen** [comment]
Yes, the test was just a demonstrator. When I look at my test I see `-lOpenDDS_QOS_XML_XSC_Handler -lACE_XML_Utils -lOpenDDS_Dcps -lTAO_BiDirGIOP -lTAO_PI -lTAO_CodecFactory -lTAO_PortableServer -lTAO_AnyTypeCode --lTAO -lACE`

In my test I get a lot of projects pulled in through the base MPC files which I all not need. Looks there is currently not a way in MPC to specific internal libs the user doesn't have to link with when doing a dynamic build, but these have to be linked in case of a static build.

**Comment by mitza-oci** [comment]
What is the problem with linking to libraries that must be present on the system?

**Comment by jwillemsen** [comment]
It looks to me as user strange that I am using OpenDDS but have to link my code with all kinds of libraries that are used internally by OpenDDS. 

An internal added library is now exposed to all users. At the moment OpenDDS could for example be compiled without inforepo the requirement for PortableServer gets removed, but that means the user has to update his project files.

**Comment by mitza-oci** [comment]
I don't know if we have a way to move forward on this issue.  The way libraries are used is "by design."  If you want to propose something different that still works for static builds, I'm sure we can discuss that.

The use cases that were mentioned in the prior comment all stem from changes to the OpenDDS repository.  When the OpenDDS repository changes, downstream projects needs to update to match.  This is easily done with MPC or CMake or not-so-easily done by manually maintaining builds.

**Comment by jwillemsen** [comment]
I am thinking about the direction that the mpc base projects that are used by the users of OpenDDS (unit tests or external projects) only get the required after and the library is only listed as static_lib. That way any internal library is not linked with user code

**Comment by jwillemsen** [comment]
So a different base project that is used for users of OpenDDS as for the core libraries itself. I will try to make a proposal for xercesc first, not sure whether I get time for it the coming weeks before New Year

**Comment by iguessthislldo** [comment]
> So a different base project that is used for users of OpenDDS as for the core libraries itself.

If it works with static like Adam if worried about, I would welcome this change.

**Comment by jwillemsen** [comment]
CMake seems to support PRIVATE (https://cmake.org/cmake/help/latest/command/target_link_libraries.html#libraries-for-a-target-and-or-its-dependents) for this.
